invenio_app_ils/records/api.py

```diff
@@ -607,57 +615,77 @@ class DocumentRequest(IlsRecord):
+        "{document_request_pid}/patron"
+    )
 
     @classmethod
-    def _validate_fulfilled(cls, document_pid):
-        """Validate data for fulfilled state."""
-        # Fulfilled requests must have a document
-        if not document_pid:
+    def validate_state(cls, state, document_pid=None):
+        """Validate state."""
+        if state not in cls.STATES:
             raise DocumentRequestError(
-                "State cannot be 'FULFILLED' without a document"
+                "Invalid state: {}".format(state)
             )
-        try:
-            document = Document.get_record_by_pid(document_pid)
-            document = document.replace_refs()
-            request = document["request"]
-            if request:
-                # The document cannot already have a request
-                raise DocumentRequestError(
-                    "Document with PID {} already has a request".format(
-                        document_pid
+
+    @classmethod
+    def validate_document_pid(cls, document_pid, state):
+        """Validate data for accepted state."""
+        # Accepted requests must have a document
+        if document_pid and (not state or state != "ACCEPTED"):
+            raise DocumentRequestError(
+                "document_pid can only be provided when accepting a request"
+            )
+        if state == "ACCEPTED":
+            if document_pid:
+                try:
+                    document = Document.get_record_by_pid(document_pid)
+                    document = document.replace_refs()
+                    request = document["request"]
+                    if request:
+                        # The document cannot already have a request
+                        raise DocumentRequestError(
+                            "Document PID {} already has a request".format(
+                                document_pid
+                            )
+                        )
+                except PIDDoesNotExistError:
+                    # Invalid document_pid
+                    raise DocumentRequestError(
+                        "State cannot be ACCEPTED because a document with "
+                        "PID {} doesn't exist".format(document_pid)
                     )
+            else:
+                raise DocumentRequestError(
+                    "State cannot be ACCEPTED without a document"
                 )
-        except PIDDoesNotExistError:
-            # Invalid document_pid
+
+    @classmethod
+    def validate_rejection(cls, state, reason):
+        """Validate rejection is correct."""
+        if state == "REJECTED" and not reason:
             raise DocumentRequestError(
-                "State cannot be 'FULFILLED' because a document with "
-                "PID {} doesn't exist".format(document_pid)
+                "Need to provide a reason when rejecting a request"
             )
 
     @classmethod
-    def validate_create(cls, data):
+    def validate_document_request(cls, data):
```

Comment:
If we are about to have this kind of custom validation in general isn't it a better idea to create something like a `validator.py` and a `Validator` class that has a `validate` function to isolate the validation and avoid the `api` file to grow bigger? Something like:

```suggestion
     class BaseValidator(object):
     """Common validator."""
     
     def validate(data):
     """validate data."""
    
    class DocumentRequestValidator(BaseValidator):
    
       def validate(self, data):
       """Specific validation"""
       
    
    
    
```

Reply:
Created issue to fix this: https://github.com/inveniosoftware/invenio-app-ils/issues/556